### ISSUE TITLE
how to set ecg to timestamp the local time

### ISSUE THREAD

**Issue by wilbertmatthew** [polarofficial/polar-ble-sdk#236]
***Platform your question concerns:*** 
- [ ] Android 
- [ ] iOS
- [ ] Other
- [ ] Platform is not relevant for this question

***Device:***
- [ ] Polar OH1
- [ ] Polar Verity Sense
- [ ] Polar H10
- [ ] Polar H9
- [ ] Other
- [ ] Device is not relevant for this question

**Description:**

**Comment by wilbertmatthew** [comment]
I am trying to set the local time so that the timestamps I get returned from the ECG Stream event are ECG local timestamps.

.observeOn(Schedulers.newThread())
       .subscribe(
        polarEcgData -> {                                   
        long timeStamp = polarEcgData.timeStamp;

       timeStamp is: 599623564424327296

To change this timestamp to the local timestamp I use the below call in the device connection event but no time change occurs. I can see that cal.getTimeInMillseconds() returns the local time stamp but the ecg data timestamps do not change.

Calendar cal = Calendar.getInstance();           
  m_api.setLocalTime(deviceId, cal);

**Comment by JOikarinen** [comment]
Hi @wilbertmatthew, 

`setLocalTime()` is [reactive](https://github.com/ReactiveX/RxJava) function. The call `m_api.setLocalTime(deviceId, cal);`returns you the Rx `Completable`. Just calling the `setLocalTime()` nothing happens, but once you subscribe to the  `Completable` the action is triggered. Your code shall look something like this:

```
Calendar cal = Calendar.getInstance();
m_api.setLocalTime(deviceId, cal).subscribe();
```

You may see the [example app](https://github.com/polarofficial/polar-ble-sdk/tree/master/examples/example-android/androidBleSdkTestApp) and play around with that. If I add log print to example app, like this

```
...
.subscribe(
          { polarEcgData: PolarEcgData ->
              Log.d("TESTING", "Timestamp "+ polarEcgData.timeStamp)
              for (microVolts in polarEcgData.samples) {
                  Log.d(TAG, "    yV: $microVolts")
              }
          },
...          
```

With these steps:
1. Connect to H10
2. Start ECG Stream
3. Set Time (the last button in example app)


I get following output:

```
...
2022-02-11 07:26:28.111 6439-6473/com.polar.androidblesdk D/TESTING: Timestamp 599617124858216450
2022-02-11 07:26:28.691 6439-6473/com.polar.androidblesdk D/TESTING: Timestamp 599617125419996880
2022-02-11 07:26:29.239 6439-6473/com.polar.androidblesdk D/TESTING: Timestamp 697879588139584244
2022-02-11 07:26:29.788 6439-6473/com.polar.androidblesdk D/TESTING: Timestamp 697879588701364674
...
```

**Comment by wilbertmatthew** [comment]
Thanks for your response. In my app I get the
error polar.com.sdk.api.errors.PolarServiceNotAvailable. All other
services work.

In the ECGActivity class the code I use:

Also tried setting context as Context ctx = getApplicationContext();
instead of this

api = PolarBleApiDefaultImpl.defaultImplementation(this,
PolarBleApi.ALL_FEATURES);

Calendar calendar = Calendar.getInstance();
calendar.setTime(new Date());
Log.d(TAG, "calendar: " + calendar); // is valid
Log.d(TAG, "deviceId: " + deviceId); // is valid
Log.d(TAG, "api: " + api); // is valid
api.setLocalTime(deviceId, calendar)
        .subscribe(
                () -> Log.d(TAG, "time set to device"),
                error -> Log.d(TAG, "set time failed: " + error));

error is always: polar.com.sdk.api.errors.PolarServiceNotAvailable.

**Comment by wilbertmatthew** [comment]
To update:  The settime error has changed. When I click the set time button
I get the error.

java.lang.NoClassDefFoundError: Failed resolution of:
Lprotocol/PftpRequest$PbPFtpSetLocalTimeParams;

Code below is in the button click event and button clicked after the ECG
Stream is started. ECG Streams are ok.

Also tried setting context as Context ctx = getApplicationContext();
instead of this

api = PolarBleApiDefaultImpl.defaultImplementation(this,
PolarBleApi.ALL_FEATURES);

Calendar calendar = Calendar.getInstance();
calendar.setTime(new Date());
Log.d(TAG, "calendar: " + calendar); // is valid
Log.d(TAG, "deviceId: " + deviceId); // is valid
Log.d(TAG, "api: " + api); // is valid
api.setLocalTime(deviceId, calendar)
        .subscribe(
                () -> Log.d(TAG, "time set to device"),
                error -> Log.d(TAG, "set time failed: " + error));

**Comment by JOikarinen** [comment]
@wilbertmatthew please check that [you have needed dependencies](https://github.com/polarofficial/polar-ble-sdk/blob/master/README.md#installation) set in your project .gradle build file:

```gradle
dependencies {
    ...
    implementation 'commons-io:commons-io:2.11.0' // Only needed if FEATURE_POLAR_FILE_TRANSFER used
    implementation 'com.google.protobuf:protobuf-javalite:3.17.3' // Only needed if FEATURE_POLAR_FILE_TRANSFER used
    ...
}
```

The error in the latest comment from you hints that `implementation 'com.google.protobuf:protobuf-javalite:3.17.3'` may be missing.

**Comment by JOikarinen** [comment]
Okay. This happens because the latest polar-ble-sdk.aar you have taken into use is expected to compile using [API 31](https://developer.android.com/about/versions/12) and its target is [API 31]. That changed in [release 3.2.7](https://github.com/polarofficial/polar-ble-sdk/releases/tag/3.2.7)
 
Here are your options:

1.  you upgrade your own application to target API 31. So in your module gradle file the `compileSdkVersion` and `targetSdkVersion` are upgraded to target API 31:
```
...
 compileSdkVersion 31
...
  targetSdkVersion 31
``` 

2.  You rollback to the .aar in [release 3.2.3 ](https://github.com/polarofficial/polar-ble-sdk/tree/3.2.3/polar-sdk-android/libs) which I think you were already using. 
3. Or you may just simply change only `compileSdkVersion` to target 31 on your application. With this option keep in mind that it is recommended to keep  `compileSdkVersion` and `targetSdkVersion` in same version.

**Comment by wilbertmatthew** [comment]
Thanks for your response. I am using the new .aar file, from your example,
and changed the compileSdkVersion and targetSdkVersion to target 31 in my
project. I clean the build and invalidate the cache and restart. After
those steps my project code shows that the PolarBleApiDefaultImpl symbol
cannot be resolved. I checked the .aar classes file and can see that
the PolarBleApiDefaultImpl is implemented yet the import statement (import
com.polar.sdk.api.PolarBleApiDefaultImpl) says the symbol cannot be
resolved. How can I fix this error? I have attached a screenshot of the
classes.jar contained in the version of the .aar file I am using.

**Comment by wilbertmatthew** [comment]
Thanks for your response. I am using the new .aar file, from your example,
and changed the compileSdkVersion and targetSdkVersion to target 31 in my
project. I clean the build and invalidate the cache and restart. After
those steps my project code shows that the PolarBleApiDefaultImpl symbol
cannot be resolved. I checked the .aar classes file and can see that
the PolarBleApiDefaultImpl is implemented yet the import statement (import
com.polar.sdk.api.PolarBleApiDefaultImpl) says the symbol cannot be
resolved. How can I fix this error? I have attached a screenshot of the
classes.jar contained in the version of the .aar file I am using.
